packages/core/src/domain/session/sessionCookieStore.spec.ts

```diff
@@ -0,0 +1,199 @@
+import { updateExperimentalFeatures, resetExperimentalFeatures } from '@datadog/browser-core'
+import { stubCookie } from '../../../test/specHelper'
+import {
+  SESSION_COOKIE_NAME,
+  toSessionString,
+  retrieveSession,
+  persistSession,
+  withCookieLockAccess,
+} from './sessionCookieStore'
+import { SessionState } from './sessionStore'
+
+describe('session cookie store', () => {
+  const COOKIE_OPTIONS = {}
+  let initialSession: SessionState
+  let otherSession: SessionState
+  let processSpy: jasmine.Spy<jasmine.Func>
+  let afterSpy: jasmine.Spy<jasmine.Func>
+  let cookie: ReturnType<typeof stubCookie>
+
+  type OnLockCheck = () => { currentState: SessionState; retryState: SessionState }
+
+  function lockScenario({
+    onInitialLockCheck,
+    onAcquiredLockCheck,
+    onPostProcessLockCheck,
+    onPostPersistLockCheck,
+  }: {
+    onInitialLockCheck?: OnLockCheck
+    onAcquiredLockCheck?: OnLockCheck
+    onPostProcessLockCheck?: OnLockCheck
+    onPostPersistLockCheck?: OnLockCheck
+  }) {
+    const onLockChecks = [onInitialLockCheck, onAcquiredLockCheck, onPostProcessLockCheck, onPostPersistLockCheck]
+    let currentSession: string
+    cookie.setSpy.and.callFake((session) => {
+      currentSession = session
+    })
+
+    cookie.getSpy.and.callFake(() => {
+      const currentOnLockCheck = onLockChecks.shift()
+      if (!currentOnLockCheck) {
+        return currentSession
+      }
+      cookie.getSpy.and.callThrough()
+      cookie.setSpy.and.callThrough()
+      const { currentState, retryState } = currentOnLockCheck()
+      persistSession(retryState, COOKIE_OPTIONS)
+      return `${SESSION_COOKIE_NAME}=${toSessionString(currentState)}`
```

Comment:
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** ‏This is really hard to get. What about not relying on the original cookie stub spies and keep using the `currentSession` instead?

```ts
      const currentOnLockCheck = onLockChecks.shift()
      if (currentOnLockCheck) {
        const { currentState, retryState } = currentOnLockCheck()
        currentSession = `${SESSION_COOKIE_NAME}=${toSessionString(currentState)}`
        return `${SESSION_COOKIE_NAME}=${toSessionString(retryState)}`
      } else {
        return currentSession
      }
```

Reply:
I've made a simplification based on your proposition but not sure where the `currentSession` came from on the proposed code, so let me know if it was what you had in mind.

Reply:
The `currentSession` is [the local variable you introduced to store the current session cookie](https://github.com/DataDog/browser-sdk/blob/8eb1617f3c9db8176a8abbafdec6b518f45d1548/packages/core/src/domain/session/sessionCookieStore.spec.ts#L34) . Your change does exactly the same but with a bit of indirection, it looks good to me :)